pep-0497-externals.txt

```diff
@@ -0,0 +1,359 @@
+
+Rationale
+=========
+
+Differences between environments have implications for the building, testing,
+installing and execution of software. There are two sides to this coin. First,
+in some environments, additional Python packages are needed. Secondly, in some
+environments, additional non-Python packages need to be present, such as header
+files, libraries, scripts or other binaries. The syntax proposed here allows
+the description of both sorts of requirements in a single syntax.
+
+Setuptools and pip should be modified to support the use case of installing
+the appropriate python packages as required depending on the specification.
+
+They should also support aborting an install with a useful error message if
+required dependecies are not installed in the environment. The responsibility
+for actually installing those required packages lies with the user.
+
+This would allow automatic installation of environment-specific Python
+dependencies and pave the way to improving how users go about installing
+non-Python prerequisites.
+
+Ideally, and error message will display the package manager command required
+to add the missing environment dependencies.
+
+It is expected that this feature would be commonly used, particularly in
+scientific applications which frequently have complex interactions with
+externally provided libraries. An example would be needing versions of
+numpy and scipy, both of which must have been compiled to be aware of the ATLAS
+compiled set of linear algebra libraries for performance reasons. This sounds
+esoteric but is, in fact, a routinely encountered situation which drives
+people towards using the alternative packaging for scientific python
+environments.
+
+Syntax
+======
+
+The valid context specifiers are taken from the environment markers library.
+This command says that when the operating system matches 'ubuntu', then add
+the yaml==0.3 requirement to the list.
+
+Meta: I am not a syntax wizard and will need help from others to properly specify
+the syntax as a formal grammar if that is required.
+
+Essentially, the "!" separator is used to flag key-value pairs which will allow
+two things: the specification of dependency type and the environment context
+where the rule should apply.
+
+For example, consider the line
+
+    setup_requires=['bin!fortran:dist_name=="ubuntu"']
+
+The 'bin' specifies one of "bin", "lib" or "header", representing the only
+non-Python dependency types allows. The absence of this prefix implies the the
+package name is a discoverable Python package.
+
+Following the ':' is a comma-separated list of key-operator-value pairs. These
+represent the sufficient trigger conditions to apply the rule. At this stage
+only a very limited set of allowable operators and logics are permitted. The
+goal is to provide just enough support to allow the specification of key
+dependencies for specific environments, not to become a make/build system.
+
+Implementation Concerns
+=======================
+
```

Comment:
platform abstraction needs to cover:

- how pip/setuptools should access it
- the API for using it
- who is expected to implement it
- how it gets into virtualenvironments


strawman:
import vendorplatform;
p = vendorplatform.Platform()

Vendors (distros, conda, etc) will include a vendorplatform package or module in their base Python. It must be selfcontained (e.g. it can be a .so or a .py or a package, but it can't call into other non-stdlib packages).

virtualenv should be modified to copy that into virtualenvs that it creates.

If vendorplatform can't be imported, the calling code should just not attempt to translate calls.